src/main/java/com/epam/aidial/core/controller/UploadFileController.java

```diff
@@ -29,28 +30,21 @@ protected Future<?> handle(ResourceDescription resource, boolean hasWriteAccess)
             return context.respond(HttpStatus.BAD_REQUEST, "Resource name and/or parent folders must not end with .(dot)");
         }
 
-        Promise<Void> result = Promise.promise();
+        EtagHeader etag = EtagHeader.fromRequest(context.getRequest());
         context.getRequest()
                 .setExpectMultipart(true)
                 .uploadHandler(upload -> {
                     String contentType = upload.contentType();
                     Pipe<Buffer> pipe = new PipeImpl<>(upload).endOnFailure(false);
-                    BlobWriteStream writeStream = new BlobWriteStream(
-                            proxy.getVertx(),
-                            proxy.getStorage(),
-                            resource,
-                            contentType);
-                    pipe.to(writeStream, result);
-
-                    result.future()
-                            .onSuccess(success -> context.respond(HttpStatus.OK, writeStream.getMetadata()))
-                            .onFailure(error -> {
-                                writeStream.abortUpload(error);
-                                context.respond(HttpStatus.INTERNAL_SERVER_ERROR,
-                                        "Failed to upload file by path %s/%s".formatted(resource.getBucketName(), resource.getOriginalPath()));
-                            });
+                    ResourceService resourceService = proxy.getResourceService();
+                    resourceService.putFile(resource, pipe, etag, contentType)
```

Comment:
Contains blocking call, needs to be called with vertx.executeBlocking

Reply:
Fixed.